wcfsetup/install/files/js/WCF.js

```diff
@@ -6067,24 +6084,40 @@ if (COMPILER_TARGET_DEFAULT) {
 		 * @return        integer
 		 */
 		_createUploadMatrix: function (files) {
-			if (files.length) {
-				var $uploadID = this._uploadMatrix.length;
-				this._uploadMatrix[$uploadID] = [];
-				
-				for (var $i = 0, $length = files.length; $i < $length; $i++) {
-					var $file = files[$i];
-					var $li = this._initFile($file);
+			var self = this;
+			
+			var updateMatrix = function (files) {
+				if (files.length) {
+					var $uploadID = self._uploadMatrix.length;
+					self._uploadMatrix[$uploadID] = [];
 					
-					if (!$li.hasClass('uploadFailed')) {
-						$li.data('filename', $file.name).data('internalFileID', this._internalFileID++);
-						this._uploadMatrix[$uploadID][$i] = $li;
+					for (var $i = 0, $length = files.length; $i < $length; $i++) {
+						var $file = files[$i];
+						var $li = self._initFile($file);
+						
+						if (!$li.hasClass('uploadFailed')) {
+							$li.data('filename', $file.name).data('internalFileID', self._internalFileID++);
+							self._uploadMatrix[$uploadID][$i] = $li;
+						}
 					}
+					
+					return $uploadID;
 				}
 				
-				return $uploadID;
+				return null;
+				
 			}
 			
-			return null;
+			if (files instanceof Promise) {
```

Comment:
See above.

Reply:
I wanted to keep the old code path as untouched as possible but makes sense to just return a promise as we have to handle them anyway.